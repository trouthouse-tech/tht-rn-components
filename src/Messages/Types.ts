import {GestureResponderEvent} from 'react-native';

type BaseProps = {
  name: string;
  date: Date;
  status: MessageStatusEnum;
  image?: string;
};

type BaseTogglerProps = {
  onPress: (selectedMessageStatus: MessageType) => void;
};

export enum MessageStatusEnum {
  Scheduled = 'Scheduled',
  Canceled = 'Canceled',
  Completed = 'Completed',
}

export enum MessageType {
  Sessions = 'Sessions',
  Requests = 'Requests',
}

export enum ScheduledMessageType {
  Scheduled = 'Scheduled',
  History = 'History',
}

//the first component type props
export type MessageItemProps = BaseProps & {
  messageType: MessageType;
};

export type MessageListProps = {
  messages: MessageItemProps[];
};

export type MessageTogglerProps = BaseTogglerProps & {
  currentMessageType: MessageType;
};

//the second component type props
export type ScheduledMessageItemProps = BaseProps & {
  notes: string;
  durationInHours: number;
  messageType: ScheduledMessageType;
};

export type ScheduledMessageListProps = {
  messages: ScheduledMessageItemProps[];
};
