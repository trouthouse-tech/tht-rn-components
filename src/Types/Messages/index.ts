import {BaseMessageProps} from '../Base/';

//the first component type props
export enum MessageTypeEnum {
  Sessions = 'Sessions',
  Requests = 'Requests',
}

export type MessageItemProps = BaseMessageProps & {
  messageType: MessageTypeEnum;
};

export type MessageListProps = {
  messages: MessageItemProps[];
};

export type MessageTogglerProps = {
  currentMessageType: MessageTypeEnum;
  onPress: (selectedMessageStatus: MessageTypeEnum) => void;
};

//the second component type props
