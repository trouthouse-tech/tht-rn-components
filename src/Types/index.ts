import * as BaseTypes from './Base/';
import * as MessageTypes from './Messages/';
import * as SessionTypes from './Sessions/';

export {BaseTypes, MessageTypes, SessionTypes};
