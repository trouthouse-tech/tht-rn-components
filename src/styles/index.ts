import * as Colors from './color';
import * as Fonts from './font';

export {Colors, Fonts};
