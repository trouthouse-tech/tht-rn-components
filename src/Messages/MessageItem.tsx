import React from 'react';
import moment from 'moment';

import {View, Text, Image, StyleSheet, Dimensions} from 'react-native';

import {MessageItemProps, MessageStatusEnum} from './Types';
import {Colors} from '../styles/Color';

export const MessageItem: React.FC<MessageItemProps> = (props) => {
  const {
    name,
    date,
    messageType,
    status,
    image = 'https://via.placeholder.com/150',
  } = props;
  return (
    <View style={styles.messageStyle}>
      <View>
        <Text style={styles.nameTextStyle}>{name}</Text>

        {status !== MessageStatusEnum.Completed ? (
          <View style={styles.dateTextViewStyle}>
            <Text style={styles.dateDisplayTextStyle}>
              {moment(date).format('DD/MM/YY')}
            </Text>
            <Text style={styles.dateDisplayTextStyle}>
              {moment(date).format('h:mm a')}
            </Text>
          </View>
        ) : (
          <Text style={styles.statusDisplayTextStyle}>{status}</Text>
        )}
        <Text>{messageType}</Text>
      </View>
      <Image source={{uri: image}} style={styles.messageImgStyle} />
    </View>
  );
};

const deviceHeight = Dimensions.get('window').height;
const styles = StyleSheet.create({
  messageStyle: {
    borderWidth: StyleSheet.hairlineWidth,
    borderRadius: 15,
    padding: 20,
    marginVertical: 10,
    justifyContent: 'space-between',
    flex: 1,
    flexDirection: 'row',
    alignContent: 'center',
    alignItems: 'center',
  },
  nameTextStyle: {
    color: Colors.Black,
    fontSize: 20,
    fontWeight: '700',
  },
  dateTextViewStyle: {
    flexDirection: 'row',
    justifyContent: 'flex-start',
    paddingVertical: 4,
  },
  dateDisplayTextStyle: {
    paddingHorizontal: 3,
    fontSize: 16,
    fontWeight: '600',
  },
  statusDisplayTextStyle: {
    fontSize: 16,
    fontWeight: '600',
  },
  messageImgStyle: {
    height: deviceHeight / 7,
    width: deviceHeight / 7,
    borderRadius: 200,
  },
});
