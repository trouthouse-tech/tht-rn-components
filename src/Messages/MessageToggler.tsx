import React from 'react';

import {View, Text, StyleSheet, TouchableWithoutFeedback} from 'react-native';

import {MessageTogglerProps, MessageType} from './Types';
import {Colors} from '../styles/';

export const MessageToggler: React.FC<MessageTogglerProps> = (props) => {
  const {currentMessageType, onPress} = props;

  return (
    <View>
      {Object.values(MessageType).map((key: MessageType, index: number) => (
        <TouchableWithoutFeedback
          key={index}
          onPress={() => {
            onPress(key);
          }}>
          <Text
            style={
              currentMessageType === key
                ? styles.selectedTabStyle
                : styles.tabStyle
            }>
            {MessageType[key]}
          </Text>
        </TouchableWithoutFeedback>
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  selectedTabStyle: {
    fontSize: 15,
    textDecorationLine: 'underline',
  },
  tabStyle: {
    fontSize: 15,
    color: Colors.Grey,
  },
});
