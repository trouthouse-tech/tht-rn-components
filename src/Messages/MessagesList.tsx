import React from 'react';

import {View, FlatList, StyleSheet} from 'react-native';

import {MessageTypes} from '../Types/';
import {MessageItem} from './MessageItem';

export const MessageList: React.FC<MessageTypes.MessageListProps> = ({
  messages,
}) => {
  return (
    <View style={styles.listStyle}>
      <FlatList
        keyExtractor={(item: MessageTypes.MessageItemProps, index: number) =>
          index.toString()
        }
        data={messages}
        renderItem={({item}: {item: MessageTypes.MessageItemProps}) => (
          <MessageItem
            name={item.name}
            status={item.status}
            messageType={item.messageType}
            date={item.date}
          />
        )}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  listStyle: {
    marginVertical: 10,
  },
});
