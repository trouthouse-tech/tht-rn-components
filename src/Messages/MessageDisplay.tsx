import React from 'react';

import {View, Text} from 'react-native';

import {MessageListProps, MessageType} from './Types';
import {MessageList} from './MessagesList';
import {MessageToggler} from './MessageToggler';

export const MessageDisplay: React.FC<MessageListProps> = ({messages}) => {
  const [currentMessageStatus, setcurrentMessageStatus] = React.useState<
    MessageType
  >(MessageType.Sessions);

  const toggleMessageStatus = (selectedMessageStatus: MessageType) => {
    console.log(selectedMessageStatus);
    setcurrentMessageStatus(selectedMessageStatus);
  };

  return (
    <View>
      <MessageToggler
        onPress={(selectedMessageStatus: MessageType) => {
          return toggleMessageStatus(selectedMessageStatus);
        }}
        currentMessageType={currentMessageStatus}
      />
      <MessageList
        messages={messages.filter(
          (m) => m.messageType === currentMessageStatus
        )}
      />
      <Text>{currentMessageStatus}</Text>
    </View>
  );
};
