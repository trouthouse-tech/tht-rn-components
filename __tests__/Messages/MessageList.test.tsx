import {render, cleanup} from 'react-native-testing-library';
import React from 'react';

import {MessageList} from '../../src/Messages/MessagesList';
import {MessageTypes, BaseTypes} from '../../src/Types/';

afterEach(cleanup);

describe('<MessageList />', () => {
  const messagesList: MessageTypes.MessageListProps = {
    messages: [
      {
        messageType: MessageTypes.MessageTypeEnum.Requests,
        name: 'Edwards Moses',
        status: BaseTypes.MessageStatusEnum.Completed,
        date: 1598457958,
      },
      {
        messageType: MessageTypes.MessageTypeEnum.Sessions,
        name: 'Edwards Moses',
        status: BaseTypes.MessageStatusEnum.Canceled,
        date: 1598457958,
      },
    ],
  };

  it('should match snapshot', () => {
    const rendered = render(
      <MessageList messages={messagesList.messages} />
    ).toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
