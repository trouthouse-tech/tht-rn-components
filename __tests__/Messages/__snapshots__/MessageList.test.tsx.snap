// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageList /> should match snapshot 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "date": 1598457958,
          "messageType": "Requests",
          "name": "Edwards Moses",
          "status": "Completed",
        },
        Object {
          "date": 1598457958,
          "messageType": "Sessions",
          "name": "Edwards Moses",
          "status": "Canceled",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "borderRadius": 15,
              "borderWidth": 0.5,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 10,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            }
          }
        >
          <View>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 20,
                  "fontWeight": "700",
                }
              }
            >
              Edwards Moses
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                }
              }
            >
              Completed
            </Text>
          </View>
          <Image
            source={
              Object {
                "uri": "https://via.placeholder.com/150",
              }
            }
            style={
              Object {
                "borderRadius": 200,
                "height": 190.57142857142858,
                "width": 190.57142857142858,
              }
            }
          />
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "borderRadius": 15,
              "borderWidth": 0.5,
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 10,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            }
          }
        >
          <View>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 20,
                  "fontWeight": "700",
                }
              }
            >
              Edwards Moses
            </Text>
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "paddingVertical": 4,
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 16,
                    "fontWeight": "600",
                    "paddingHorizontal": 3,
                  }
                }
              >
                26/08/20
              </Text>
              <Text
                style={
                  Object {
                    "fontSize": 16,
                    "fontWeight": "600",
                    "paddingHorizontal": 3,
                  }
                }
              >
                5:05 pm
              </Text>
            </View>
          </View>
          <Image
            source={
              Object {
                "uri": "https://via.placeholder.com/150",
              }
            }
            style={
              Object {
                "borderRadius": 200,
                "height": 190.57142857142858,
                "width": 190.57142857142858,
              }
            }
          />
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
